# `to_dataframe` functions demonstrations

Works with both normal and IO databases. Start with a normal database (stored normally, even if the origin is an IO table).

Normal databases have each edge stored in the SQLite table as a separate row.

In [1]:
import bw2data as bd
import bw2calc as bc
import bw2io as bi

In [55]:
bd.projects.set_current("to_dataframe USEEIO")

Utility function to get the US EEIO and its LCIA methods easily. This has its own biosphere flows, so we don't run `bw2setup`.

In [3]:
bi.useeio11()

13238272it [00:01, 8304141.85it/s]                                                                                                                             


Unzipping file
Importing data
Applying strategy: json_ld_allocate_datasets
Applying strategy: json_ld_get_normalized_exchange_locations
Applying strategy: json_ld_convert_unit_to_reference_unit
Applying strategy: json_ld_get_activities_list_from_rawdata
Applying strategy: json_ld_add_products_as_activities
Applying strategy: json_ld_get_normalized_exchange_units
Applying strategy: json_ld_add_activity_unit
Applying strategy: json_ld_rename_metadata_fields
Applying strategy: json_ld_location_name
Applying strategy: json_ld_remove_fields
Applying strategy: json_ld_fix_process_type
Applying strategy: json_ld_label_exchange_type
Applying strategy: json_ld_prepare_exchange_fields_for_linking
Applying strategy: add_database_name
Applying strategy: link_iterable_by_fields
Applying strategy: link_iterable_by_fields


Writing activities to SQLite3 database:


Applying strategy: normalize_units
Applied 17 strategies in 0.56 seconds
Moved 1873 biosphere flows to `self.data`
2649 datasets
162926 exchanges
0 unlinked exchanges
  
Not able to determine geocollections for all datasets. This database is not ready for regionalization.


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:04


Title: Writing activities to SQLite3 database:
  Started: 08/21/2022 21:34:26
  Finished: 08/21/2022 21:34:30
  Total time elapsed: 00:00:04
  CPU %: 97.70
  Memory %: 1.89
Created database: US EEIO 1.1
Applying strategy: json_ld_lcia_add_method_metadata
Applying strategy: json_ld_lcia_convert_to_list
Applying strategy: json_ld_lcia_set_method_metadata
Applying strategy: json_ld_lcia_reformat_cfs_as_exchanges
Applying strategy: normalize_units
Applied 5 strategies in 0.01 seconds
19 methods
4511 cfs
0 unlinked cfs
Wrote 19 LCIA methods with 4511 characterization factors


Pick a product and an activity node at random.

In [48]:
product = next(node for node in bd.Database("US EEIO 1.1") if node['type'] == 'product')
activity = next(node for node in bd.Database("US EEIO 1.1") if node['type'] == 'process')
product, activity

('Scientific research and development' (, United States, ('54: Professional, Scientific, and Technical Services', '5417: Scientific Research and Development Services')),
 'Other retail' (USD, United States, None))

The first dataframe is all the nodes (processes or activities) in the given database:

In [49]:
df = bd.Database("US EEIO 1.1").nodes_to_dataframe()
df

CAS number                     categories classifications  \
1175                      (water, unspecified)             NaN   
1745             (air, low population density)             NaN   
1948                       (soil, groundwater)             NaN   
1174                      (water, unspecified)             NaN   
2300                       (soil, groundwater)             NaN   
...         ...                            ...             ...   
1088                        (air, unspecified)             NaN   
222   7440622.0             (soil, industrial)             NaN   
1146  7440622.0           (water, unspecified)             NaN   
2373  7440622.0             (air, unspecified)             NaN   
1863                        (air, unspecified)             NaN   

                                      code     database description dqEntry  \
1175  2ee4697d-b7f4-362b-86a4-94b644699500  US EEIO 1.1         NaN     NaN   
1745  6ca23b5d-83dc-3b02-bf39-8eabf9d41151  US EEIO 1.1         NaN     NaN   
1948  5b98f875-8d1c-3549-a7df-28d7d90e7ccb  US EEIO 1.1         NaN     NaN   
1174  93086e32-c013-3e34-a074-4760c72fe775  US EEIO 1.1         NaN     NaN   
2300  b8889098-9a89-35c6-b226-07cb66c83217  US EEIO 1.1         NaN     NaN   
...                                    ...          ...         ...     ...   
1088  d9a5b786-d06c-44af-a088-b070aa605d9b  US EEIO 1.1         NaN     NaN   
222   1a5850a0-0069-4b73-bb91-7a61e8d45ae5  US EEIO 1.1         NaN     NaN   
1146  63e8256e-8549-11e0-9d78-0800200c9a66  US EEIO 1.1         NaN     NaN   
2373  591b0a62-8064-4697-86ed-47bfa1f8b5e6  US EEIO 1.1         NaN     NaN   
1863  1d444a45-70c4-4718-8427-1c0fff91f88e  US EEIO 1.1         NaN     NaN   

     dqSystem exchangeDqSystem filename    id location modified  \
1175      NaN              NaN      NaN  1134      NaN      NaN   
1745      NaN              NaN      NaN  1640      NaN      NaN   
1948      NaN              NaN      NaN  1427      NaN      NaN   
1174      NaN              NaN      NaN  1886      NaN      NaN   
2300      NaN              NaN      NaN  2160      NaN      NaN   
...       ...              ...      ...   ...      ...      ...   
1088      NaN              NaN      NaN  2370      NaN      NaN   
222       NaN              NaN      NaN   971      NaN      NaN   
1146      NaN              NaN      NaN  1577      NaN      NaN   
2373      NaN              NaN      NaN  1407      NaN      NaN   
1863      NaN              NaN      NaN  1000      NaN      NaN   

                                                   name processDocumentation  \
1175  (2,4-DICHLOROPHENOXY)ACETIC ACID COMPD. WITH 2...                  NaN   
1745  (2,4-DICHLOROPHENOXY)ACETIC ACID COMPD. WITH 2...                  NaN   
1948  (2,4-DICHLOROPHENOXY)ACETIC ACID COMPD. WITH 2...                  NaN   
1174  (4-CHLORO-2-METHYLPHENOXY)ACETIC ACID COMPD. W...                  NaN   
2300  (4-CHLORO-2-METHYLPHENOXY)ACETIC ACID COMPD. W...                  NaN   
...                                                 ...                  ...   
1088                                        trifluralin                  NaN   
222                                            vanadium                  NaN   
1146                                           vanadium                  NaN   
2373                                           vanadium                  NaN   
1863                                      vinyl acetate                  NaN   

          type unit version  
1175  emission          NaN  
1745  emission          NaN  
1948  emission          NaN  
1174  emission          NaN  
2300  emission          NaN  
...        ...  ...     ...  
1088  emission          NaN  
222   emission          NaN  
1146  emission          NaN  
2373  emission          NaN  
1863  emission          NaN  

[2649 rows x 18 columns]

The columns come from the data attributes stored on the nodes. If one node has the attribute, it is added as a column. You can control which columns get returned, and how they are sorted, see the docstring.

This is a normal dataframe, so you can filter it, add or remove columns, and sort as desired.

In [50]:
df.columns

Index(['CAS number', 'categories', 'classifications', 'code', 'database',
       'description', 'dqEntry', 'dqSystem', 'exchangeDqSystem', 'filename',
       'id', 'location', 'modified', 'name', 'processDocumentation', 'type',
       'unit', 'version'],
      dtype='object')

We can also list all the edges (exchanges) as a dataframe. This is normally too much information, and can take a bit of time to produce, but can be useful.

In [51]:
df = bd.Database("US EEIO 1.1").edges_to_dataframe()
df

Getting activity data


100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2649/2649 [00:00<00:00, 368893.76it/s]


Adding exchange data to activities


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 162926/162926 [00:02<00:00, 61446.10it/s]


Filling out exchange data


100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2649/2649 [00:00<00:00, 180277.97it/s]


Creating DataFrame
Compressing DataFrame


target_id target_database                           target_code  \
0               1     US EEIO 1.1  01624075-b520-3826-bd73-2068f7aa24e7   
1               1     US EEIO 1.1  01624075-b520-3826-bd73-2068f7aa24e7   
2               1     US EEIO 1.1  01624075-b520-3826-bd73-2068f7aa24e7   
3               1     US EEIO 1.1  01624075-b520-3826-bd73-2068f7aa24e7   
4               1     US EEIO 1.1  01624075-b520-3826-bd73-2068f7aa24e7   
...           ...             ...                                   ...   
162921        388     US EEIO 1.1  fe5971a7-2610-32ca-8193-a94873de130c   
162922        388     US EEIO 1.1  fe5971a7-2610-32ca-8193-a94873de130c   
162923        388     US EEIO 1.1  fe5971a7-2610-32ca-8193-a94873de130c   
162924        388     US EEIO 1.1  fe5971a7-2610-32ca-8193-a94873de130c   
162925        388     US EEIO 1.1  fe5971a7-2610-32ca-8193-a94873de130c   

                                              target_name  \
0                            Frozen food; at manufacturer   
1                            Frozen food; at manufacturer   
2                            Frozen food; at manufacturer   
3                            Frozen food; at manufacturer   
4                            Frozen food; at manufacturer   
...                                                   ...   
162921  Automatic controls for HVAC and refrigeration ...   
162922  Automatic controls for HVAC and refrigeration ...   
162923  Automatic controls for HVAC and refrigeration ...   
162924  Automatic controls for HVAC and refrigeration ...   
162925  Automatic controls for HVAC and refrigeration ...   

       target_reference_product target_location target_unit target_type  \
0                           NaN   United States         USD     process   
1                           NaN   United States         USD     process   
2                           NaN   United States         USD     process   
3                           NaN   United States         USD     process   
4                           NaN   United States         USD     process   
...                         ...             ...         ...         ...   
162921                      NaN   United States         USD     process   
162922                      NaN   United States         USD     process   
162923                      NaN   United States         USD     process   
162924                      NaN   United States         USD     process   
162925                      NaN   United States         USD     process   

        source_id source_database                           source_code  \
0             459     US EEIO 1.1  26bed504-3f97-3a2b-aa83-ffbe94f3b371   
1             437     US EEIO 1.1  1bafcbbb-dbe0-338d-b9c1-8c355426cbef   
2            1020     US EEIO 1.1  20185046-64bb-4c09-a8e7-e8a9e144ca98   
3            1728     US EEIO 1.1  7ae398b3-8532-11e0-9d78-0800200c9a66   
4             408     US EEIO 1.1  0bb0108f-c486-32b3-b059-e0c6c8380571   
...           ...             ...                                   ...   
162921       1426     US EEIO 1.1  5b2a19b9-1243-44ae-b76c-c0d92159d5d6   
162922       2622     US EEIO 1.1  fd7aa71c-508c-480d-81a6-8052aad92646   
162923       1462     US EEIO 1.1  5fd672a0-cb68-39e6-88dc-db1a9281c57b   
162924        589     US EEIO 1.1  7e5c6ee1-a47e-3afd-9278-dcb5e1ee50b5   
162925       1783     US EEIO 1.1  81e17d9a-3e0b-414e-9b2e-eca7e5b3bd4a   

                                              source_name source_product  \
0                            Frozen food; at manufacturer            NaN   
1                  State and local government enterprises            NaN   
2                                     Dinitrogen monoxide            NaN   
3                                         ethylene glycol            NaN   
4       Tobacco, cotton, sugarcane, peanuts, sugar bee...            NaN   
...                                                   ...            ...   
162921                                  

Now we have standard column labels. as these are directed edges, they have a source and a target. Most of the columns should be self-explanatory. Note that we differentiate between `'target_reference_product'` and `'source_product'`, and only provide the `categories` on the `source`.

In [52]:
df.columns

Index(['target_id', 'target_database', 'target_code', 'target_name',
       'target_reference_product', 'target_location', 'target_unit',
       'target_type', 'source_id', 'source_database', 'source_code',
       'source_name', 'source_product', 'source_location', 'source_unit',
       'source_categories', 'edge_amount', 'edge_type'],
      dtype='object')

If you want to add or remove columns, you can pass in an iterable of formatting functions. These functions must satisfy the following rules:

* The take the keyword arguments `node`, `edge`, and `row`.
* They modify the dictionary `row` in place. Any return value is ignored.
* `node` and `edge` are dictionaries following the [wurst internal format](https://wurst.readthedocs.io/#internal-data-format). `node` is the target, and `edge` is both attributes of the edge and of the source.

Here is a simple example:

In [61]:
def remove_target_database(node, edge, row):
    del row['target_database']
    
def food_sector(node, edge, row):
    row['is_food'] = 'food' in edge['name'].lower()

In [63]:
df = bd.Database("US EEIO 1.1").edges_to_dataframe(formatters=[remove_target_database, food_sector])
df

Getting activity data


100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2649/2649 [00:00<00:00, 268355.22it/s]


Adding exchange data to activities


100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 162926/162926 [00:02<00:00, 64784.46it/s]


Filling out exchange data


100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2649/2649 [00:00<00:00, 167645.59it/s]


Creating DataFrame
Compressing DataFrame


target_id                           target_code  \
0               1  01624075-b520-3826-bd73-2068f7aa24e7   
1               1  01624075-b520-3826-bd73-2068f7aa24e7   
2               1  01624075-b520-3826-bd73-2068f7aa24e7   
3               1  01624075-b520-3826-bd73-2068f7aa24e7   
4               1  01624075-b520-3826-bd73-2068f7aa24e7   
...           ...                                   ...   
162921        388  fe5971a7-2610-32ca-8193-a94873de130c   
162922        388  fe5971a7-2610-32ca-8193-a94873de130c   
162923        388  fe5971a7-2610-32ca-8193-a94873de130c   
162924        388  fe5971a7-2610-32ca-8193-a94873de130c   
162925        388  fe5971a7-2610-32ca-8193-a94873de130c   

                                              target_name  \
0                            Frozen food; at manufacturer   
1                            Frozen food; at manufacturer   
2                            Frozen food; at manufacturer   
3                            Frozen food; at manufacturer   
4                            Frozen food; at manufacturer   
...                                                   ...   
162921  Automatic controls for HVAC and refrigeration ...   
162922  Automatic controls for HVAC and refrigeration ...   
162923  Automatic controls for HVAC and refrigeration ...   
162924  Automatic controls for HVAC and refrigeration ...   
162925  Automatic controls for HVAC and refrigeration ...   

       target_reference_product target_location target_unit target_type  \
0                           NaN   United States         USD     process   
1                           NaN   United States         USD     process   
2                           NaN   United States         USD     process   
3                           NaN   United States         USD     process   
4                           NaN   United States         USD     process   
...                         ...             ...         ...         ...   
162921                      NaN   United States         USD     process   
162922                      NaN   United States         USD     process   
162923                      NaN   United States         USD     process   
162924                      NaN   United States         USD     process   
162925                      NaN   United States         USD     process   

        source_id source_database                           source_code  \
0             459     US EEIO 1.1  26bed504-3f97-3a2b-aa83-ffbe94f3b371   
1             437     US EEIO 1.1  1bafcbbb-dbe0-338d-b9c1-8c355426cbef   
2            1020     US EEIO 1.1  20185046-64bb-4c09-a8e7-e8a9e144ca98   
3            1728     US EEIO 1.1  7ae398b3-8532-11e0-9d78-0800200c9a66   
4             408     US EEIO 1.1  0bb0108f-c486-32b3-b059-e0c6c8380571   
...           ...             ...                                   ...   
162921       1426     US EEIO 1.1  5b2a19b9-1243-44ae-b76c-c0d92159d5d6   
162922       2622     US EEIO 1.1  fd7aa71c-508c-480d-81a6-8052aad92646   
162923       1462     US EEIO 1.1  5fd672a0-cb68-39e6-88dc-db1a9281c57b   
162924        589     US EEIO 1.1  7e5c6ee1-a47e-3afd-9278-dcb5e1ee50b5   
162925       1783     US EEIO 1.1  81e17d9a-3e0b-414e-9b2e-eca7e5b3bd4a   

                                              source_name source_product  \
0                            Frozen food; at manufacturer            NaN   
1                  State and local government enterprises            NaN   
2                                     Dinitrogen monoxide            NaN   
3                                         ethylene glycol            NaN   
4       Tobacco, cotton, sugarcane, peanuts, sugar bee...            NaN   
...                                                   ...            ...   
162921                                   2-METHOXYETHANOL            NaN   
162922                                     sulfur dioxide            NaN   
162923                    2,4,5,2',5'-PENTACHLOROBIPHENYL            NaN   
1629

In the case of `target_name`, the dataframe has more than 150.000 rows, but only 388 unique values.

You can skip the conversion to categorical columns by passing `categorical=False`.

To save on memory, we turn some columns into categorical columns, where each unique value is only stored once.

In [53]:
df.dtypes

target_id                      int64
target_database             category
target_code                   object
target_name                 category
target_reference_product    category
target_location             category
target_unit                 category
target_type                 category
source_id                      int64
source_database             category
source_code                 category
source_name                 category
source_product              category
source_location             category
source_unit                 category
source_categories           category
edge_amount                  float64
edge_type                   category
dtype: object

In [54]:
df['target_name']

0                              Frozen food; at manufacturer
1                              Frozen food; at manufacturer
2                              Frozen food; at manufacturer
3                              Frozen food; at manufacturer
4                              Frozen food; at manufacturer
                                ...                        
162921    Automatic controls for HVAC and refrigeration ...
162922    Automatic controls for HVAC and refrigeration ...
162923    Automatic controls for HVAC and refrigeration ...
162924    Automatic controls for HVAC and refrigeration ...
162925    Automatic controls for HVAC and refrigeration ...
Name: target_name, Length: 162926, dtype: category
Categories (388, object): ['Abrasive products; at manufacturer', 'Accounting, tax preparation, bookkeeping, and..., 'Adhesives; at manufacturer', 'Advertising and public relations', ..., 'Wiring devices; at manufacturer', 'Wood kitchen cabinets and countertops; at man..., 'Wood pulp; at m

We can also get a dataframe of the edges for a specific node. Here we get all edges, but you can filter this further with the edge constructors `.production()`, `.technosphere()`, and `.biosphere()`.

In [13]:
df = activity.exchanges().to_dataframe()
df

target_id target_database                           target_code  \
0          373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
1          373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
2          373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
3          373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
4          373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
..         ...             ...                                   ...   
224        373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
225        373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
226        373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
227        373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   
228        373     US EEIO 1.1  ee44f5fb-fc51-3071-a4b0-dc63b7e932bf   

                                           target_name  \
0    Other computer related services, including fac...   
1    Other computer related services, including fac...   
2    Other computer related services, including fac...   
3    Other computer related services, including fac...   
4    Other computer related services, including fac...   
..                                                 ...   
224  Other computer related services, including fac...   
225  Other computer related services, including fac...   
226  Other computer related services, including fac...   
227  Other computer related services, including fac...   
228  Other computer related services, including fac...   

    target_reference_product target_location target_unit target_type  \
0                        NaN   United States         USD     process   
1                        NaN   United States         USD     process   
2                        NaN   United States         USD     process   
3                        NaN   United States         USD     process   
4                        NaN   United States         USD     process   
..                       ...             ...         ...         ...   
224                      NaN   United States         USD     process   
225                      NaN   United States         USD     process   
226                      NaN   United States         USD     process   
227                      NaN   United States         USD     process   
228                      NaN   United States         USD     process   

     source_id source_database                           source_code  \
0          640     US EEIO 1.1  acf5099d-f095-3bdc-9efe-da7e3b535d40   
1         1832     US EEIO 1.1  88ef28f1-cfd5-44a0-ac34-01acf2db84a0   
2          464     US EEIO 1.1  2b8a18a4-1b89-3625-9263-bf9e8b6ebe1b   
3          438     US EEIO 1.1  1c64f0cb-bbab-3d07-9e38-8aff7b6746f0   
4          426     US EEIO 1.1  169c4fbb-c518-39e0-8344-3162382887c2   
..         ...             ...                                   ...   
224       1072     US EEIO 1.1  281e1011-f121-4d5b-ad4d-22d562b2c2cd   
225       1726     US EEIO 1.1  7ae371aa-8532-11e0-9d78-0800200c9a66   
226       1020     US EEIO 1.1  20185046-64bb-4c09-a8e7-e8a9e144ca98   
227       1181     US EEIO 1.1  37880630-22b3-3c88-b701-4abde6f0b9a8   
228        652     US EEIO 1.1  b53486ed-b7d6-3b78-a225-17c9dcdbffcf   

                                           source_name source_product  \
0    Other computer related services, including fac...            NaN   
1                                      benzene, ethyl-            NaN   
2                               Household goods repair            NaN   
3       Relay and industrial controls; at manufacturer            NaN   
4    Environmental and other technical consulting s...            NaN   
..                                                 ...            ...   
224                      Ethane, pentafluoro-, HFC-125            NaN   
225                                     Carbon dioxide            NaN   
226                                Dinitrogen monoxide  

Same columns as before.

In [14]:
df.columns

Index(['target_id', 'target_database', 'target_code', 'target_name',
       'target_reference_product', 'target_location', 'target_unit',
       'target_type', 'source_id', 'source_database', 'source_code',
       'source_name', 'source_product', 'source_location', 'source_unit',
       'source_categories', 'edge_amount', 'edge_type'],
      dtype='object')

We can also get dataframes for LCA calculation results.

In [18]:
lca = bc.LCA({product: 1}, method=('Impact Potential', 'HRSP'))
lca.lci()
lca.lcia()

By default, this method looks at the `characterized_inventory` matrix, and sorts by the top 200 values (using absolute value).

In [23]:
df = lca.to_dataframe()
df

row_index  col_index        amount  row_id  col_id row_database  \
0         1043          8  3.779723e-04    1822       9  US EEIO 1.1   
1          257          8  4.411418e-05    1035       9  US EEIO 1.1   
2           55          8  4.653532e-07     832       9  US EEIO 1.1   
3         1843          8  6.529780e-08    2622       9  US EEIO 1.1   
4         1445          8  3.981023e-06    2224       9  US EEIO 1.1   
..         ...        ...           ...     ...     ...          ...   
195        257        332  5.718976e-08    1035     333  US EEIO 1.1   
196        257        119  5.238594e-08    1035     120  US EEIO 1.1   
197        257        269  5.219094e-08    1035     270  US EEIO 1.1   
198        257        197  4.989681e-08    1035     198  US EEIO 1.1   
199        257        135  4.737250e-08    1035     136  US EEIO 1.1   

                                 row_code                row_name  \
0    87883a4e-1e3e-4c9d-90c0-f1bea36f8014                 ammonia   
1    21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   
2    08a91e70-3ddc-11dd-91be-0050c2490048   particulates, < 10 um   
3    fd7aa71c-508c-480d-81a6-8052aad92646          sulfur dioxide   
4    c1b91234-6f24-417b-8309-46111d09c457         nitrogen oxides   
..                                    ...                     ...   
195  21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   
196  21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   
197  21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   
198  21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   
199  21e46cb8-6233-4c99-bac3-c41d2ab99498  particulates, < 2.5 um   

    row_location row_unit  row_type    row_categories row_product  \
0           None           emission  air::unspecified        None   
1           None           emission  air::unspecified        None   
2           None           emission  air::unspecified        None   
3           None           emission  air::unspecified        None   
4           None           emission  air::unspecified        None   
..           ...      ...       ...               ...         ...   
195         None           emission  air::unspecified        None   
196         None           emission  air::unspecified        None   
197         None           emission  air::unspecified        None   
198         None           emission  air::unspecified        None   
199         None           emission  air::unspecified        None   

    col_database                              col_code  \
0    US EEIO 1.1  04ee2e71-af3b-39f3-8e69-bcae6a2d70d8   
1    US EEIO 1.1  04ee2e71-af3b-39f3-8e69-bcae6a2d70d8   
2    US EEIO 1.1  04ee2e71-af3b-39f3-8e69-bcae6a2d70d8   
3    US EEIO 1.1  04ee2e71-af3b-39f3-8e69-bcae6a2d70d8   
4    US EEIO 1.1  04ee2e71-af3b-39f3-8e69-bcae6a2d70d8   
..           ...                                   ...   
195  US EEIO 1.1  d5b7c244-8d46-3ede-9659-45986f6c6fb5   
196  US EEIO 1.1  4775afeb-8e11-3ea2-92b0-4f1c41952703   
197  US EEIO 1.1  ad81ce2a-e3c5-3695-a2ef-812cd8b79dd3   
198  US EEIO 1.1  777544f7-e9cc-3593-8d70-9a018d2a87e2   
199  US EEIO 1.1  50cee5d2-6df7-3d5f-b941-5064148172c0   

                                              col_name   col_location  \
0                                              Dairies  United States   
1                                              Dairies  United States   
2                                              Dairies  United States   
3                                              Dairies  United States   
4                                              Dairies  United States   
..                                                 ...            ...   
195             Cutlery and handtools; at manufacturer  United States   
196  Other secondary nonferrous metal products; at ...  United States   
197            Other plastic products; at manufacturer  United States   
198  Metal coatings, engravings, and heat treatment... 

The columns labels are a bit different, as we don't have target and source but instead matrix rows and columns. The meaning of these rows and columns changes from matrix to matrix. The same pattern with `'row_product'`, `'col_reference_product'`, and `'row_categories'` applies though.

In [24]:
df.columns

Index(['row_index', 'col_index', 'amount', 'row_id', 'col_id', 'row_database',
       'row_code', 'row_name', 'row_location', 'row_unit', 'row_type',
       'row_categories', 'row_product', 'col_database', 'col_code', 'col_name',
       'col_location', 'col_unit', 'col_type', 'col_reference_product'],
      dtype='object')

We can get dataframes for any matrix. In standard LCA, the matrices are:

* inventory
* technosphere_matrix
* biosphere_matrix
* characterization_matrix
* characterized_inventory

Regionalization adds more matrices. Note that for other matrices you will need to specify the row and column mapping dictionaries, see the docstring.

In [22]:
lca.to_dataframe(matrix_label='biosphere_matrix')

row_index  col_index      amount  row_id  col_id row_database  \
0          232         18  372.440552    1010      19  US EEIO 1.1   
1          947         18    0.174231    1726      19  US EEIO 1.1   
2          232         75    0.283722    1010      76  US EEIO 1.1   
3          947         75    0.116804    1726      76  US EEIO 1.1   
4         1388         95   96.912102    2167      96  US EEIO 1.1   
..         ...        ...         ...     ...     ...          ...   
195        947        121    0.092894    1726     122  US EEIO 1.1   
196        947         39    0.092228    1726      40  US EEIO 1.1   
197        947        191    0.090496    1726     192  US EEIO 1.1   
198        947        340    0.090194    1726     341  US EEIO 1.1   
199       1130        151    0.153748    1909     152  US EEIO 1.1   

                                 row_code                     row_name  \
0    1ece2361-87e0-355c-a702-ff268570ca3e                         Coal   
1    7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
2    1ece2361-87e0-355c-a702-ff268570ca3e                         Coal   
3    7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
4    b91d0527-9a01-4a86-b420-c62b70629ba4           Occupation, forest   
..                                    ...                          ...   
195  7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
196  7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
197  7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
198  7ae371aa-8532-11e0-9d78-0800200c9a66               Carbon dioxide   
199  97c598f0-4d1d-4f9f-906e-097347168153  Occupation, industrial area   

    row_location row_unit  row_type       row_categories row_product  \
0           None           emission  resource::in ground        None   
1           None           emission     air::unspecified        None   
2           None           emission  resource::in ground        None   
3           None           emission     air::unspecified        None   
4           None           emission       resource::land        None   
..           ...      ...       ...                  ...         ...   
195         None           emission     air::unspecified        None   
196         None           emission     air::unspecified        None   
197         None           emission     air::unspecified        None   
198         None           emission     air::unspecified        None   
199         None           emission       resource::land        None   

    col_database                              col_code  \
0    US EEIO 1.1  08f1c4b8-03f9-360c-87be-31ad6c778da5   
1    US EEIO 1.1  08f1c4b8-03f9-360c-87be-31ad6c778da5   
2    US EEIO 1.1  2bf3d179-abc9-3d6f-983f-f2de03471649   
3    US EEIO 1.1  2bf3d179-abc9-3d6f-983f-f2de03471649   
4    US EEIO 1.1  392eb1e3-3cd1-34c7-948c-c177114e8d20   
..           ...                                   ...   
195  US EEIO 1.1  48b2f105-8ae0-36e2-ad0a-cf7101ab8f4c   
196  US EEIO 1.1  15615fdc-2456-3e6c-bd24-9ed9a13d2599   
197  US EEIO 1.1  7452907b-74cc-3106-aaaa-5560c0645af6   
198  US EEIO 1.1  d920c723-5594-34a5-8b55-0cbe517c2f9f   
199  US EEIO 1.1  59560ccf-f607-3df2-9e6d-46400e9b8f74   

                                        col_name   col_location col_unit  \
0                                  Coal; at mine  United States      USD   
1                                  Coal; at mine  United States      USD   
2            Other support activities for mining  United States      USD   
3            Other support activities for mining  United States      USD   
4      Timber and raw forest products; at forest  United States      USD   
..                                           ...            ...      ...   
195  Residential building repair and maintanence  United States      USD   
196                        Health care buildings  United States      USD   
197            Flours and mal

We can do the same for IO databases. IO databases are stored in a different way, with numeric values only stored as Numpy arrays, but the functions work the same. This wasn't trivial, you can see the [code in the iotable backend](https://github.com/brightway-lca/brightway2-data/tree/main/bw2data/backends/iotable), but it's awesome that it works!

In [27]:
bd.projects.set_current("to_dataframe Exiobase 3.8.1 monetary")

In [28]:
bi.bw2setup()

Writing activities to SQLite3 database:


Creating default biosphere

Applying strategy: normalize_units
Applying strategy: drop_unspecified_subcategories
Applying strategy: ensure_categories_are_tuples
Applied 3 strategies in 0.00 seconds


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


Title: Writing activities to SQLite3 database:
  Started: 08/21/2022 21:44:30
  Finished: 08/21/2022 21:44:30
  Total time elapsed: 00:00:00
  CPU %: 103.10
  Memory %: 1.73
Created database: biosphere3
Creating default LCIA methods

Applying strategy: normalize_units
Applying strategy: set_biosphere_type
Applying strategy: fix_ecoinvent_38_lcia_implementation
Applying strategy: drop_unspecified_subcategories
Applying strategy: link_iterable_by_fields
Applied 5 strategies in 0.57 seconds
Wrote 975 LCIA methods with 254388 characterization factors
Creating core data migrations



Experimental function to make EXIOBASE imports easier :)

In [29]:
bi.exiobase_monetary(year=2022)

754188288it [02:04, 6047601.29it/s]                                                                                                                            
Writing activities to SQLite3 database:


Not able to determine geocollections for all datasets. This database is not ready for regionalization.


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


Title: Writing activities to SQLite3 database:
  Started: 08/21/2022 21:49:21
  Finished: 08/21/2022 21:49:21
  Total time elapsed: 00:00:00
  CPU %: 106.90
  Memory %: 2.96


Writing activities to SQLite3 database:


Created new database for EXIOBASE-specific biosphere flows: EXIOBASE 3.8.1 2022 monetary biosphere


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


Title: Writing activities to SQLite3 database:
  Started: 08/21/2022 21:49:23
  Finished: 08/21/2022 21:49:23
  Total time elapsed: 00:00:00
  CPU %: 100.90
  Memory %: 2.96
Created database of EXIOBASE activity metadata
Patched 74 LCIA methods with unit 'kg CO2-Eq'
Patching LCIA methods with EXIOBASE flows
Starting IO table write
Adding technosphere matrix


7988it [01:03, 125.03it/s]


Adding biosphere matrix


1114it [00:02, 408.37it/s]


Finalizing serialization
Created database EXIOBASE 3.8.1 2022 monetary. Cleaned up temporary downloads.


This is an instance of `IOTableBackend`.

In [37]:
bd.Database("EXIOBASE 3.8.1 2022 monetary")

Brightway2 IOTableBackend: EXIOBASE 3.8.1 2022 monetary

We can get all the nodes:

In [31]:
df = bd.Database("EXIOBASE 3.8.1 2022 monetary").nodes_to_dataframe()
df

code  \
3231  Activities auxiliary to financial intermediati...   
850   Activities auxiliary to financial intermediati...   
3541  Activities auxiliary to financial intermediati...   
5659  Activities auxiliary to financial intermediati...   
563   Activities auxiliary to financial intermediati...   
...                                                 ...   
1564                         Wool, silk-worm cocoons|WE   
384                          Wool, silk-worm cocoons|WF   
95                           Wool, silk-worm cocoons|WL   
1827                         Wool, silk-worm cocoons|WM   
246                          Wool, silk-worm cocoons|ZA   

                          database     id location  \
3231  EXIOBASE 3.8.1 2022 monetary   5670       AT   
850   EXIOBASE 3.8.1 2022 monetary  11701       AU   
3541  EXIOBASE 3.8.1 2022 monetary   5833       BE   
5659  EXIOBASE 3.8.1 2022 monetary   5996       BG   
563   EXIOBASE 3.8.1 2022 monetary  11049       BR   
...                            ...    ...      ...   
1564  EXIOBASE 3.8.1 2022 monetary  13053       WE   
384   EXIOBASE 3.8.1 2022 monetary  13216       WF   
95    EXIOBASE 3.8.1 2022 monetary  12890       WL   
1827  EXIOBASE 3.8.1 2022 monetary  13379       WM   
246   EXIOBASE 3.8.1 2022 monetary  12564       ZA   

                                                  name  production volume  \
3231  Activities auxiliary to financial intermediation        4841.805237   
850   Activities auxiliary to financial intermediation       57342.932946   
3541  Activities auxiliary to financial intermediation       20947.253461   
5659  Activities auxiliary to financial intermediation         325.803300   
563   Activities auxiliary to financial intermediation       22782.277847   
...                                                ...                ...   
1564                           Wool, silk-worm cocoons         112.305430   
384                            Wool, silk-worm cocoons        1439.132549   
95                             Wool, silk-worm cocoons         334.831769   
1827                           Wool, silk-worm cocoons         464.853941   
246                            Wool, silk-worm cocoons          51.700928   

                                     reference product     stam       unit  
3231  Activities auxiliary to financial intermediation     None  million €  
850   Activities auxiliary to financial intermediation     None  million €  
3541  Activities auxiliary to financial intermediation     None  million €  
5659  Activities auxiliary to financial intermediation     None  million €  
563   Activities auxiliary to financial intermediation     None  million €  
...                                                ...      ...        ...  
1564                           Wool, silk-worm cocoons  Animals  million €  
384                            Wool, silk-worm cocoons  Animals  million €  
95                             Wool, silk-worm cocoons  Animals  million €  
1827                           Wool, silk-worm cocoons  Animals  million €  
246                            Wool, silk-worm cocoons  Animals  million €  

[7987 rows x 9 columns]

Columns come from the attributes stored on the nodes:

In [32]:
df.columns

Index(['code', 'database', 'id', 'location', 'name', 'production volume',
       'reference product', 'stam', 'unit'],
      dtype='object')

We can also get all 39 million edges as a single dataframe, if you really want to. We use categorical columns and other tricks during construction to reduce memory usage. Same columns as before. **However**, conversion to categorical columns always happens, and you can't specify custom formatters.

In [33]:
df = bd.Database("EXIOBASE 3.8.1 2022 monetary").edges_to_dataframe()
df

Retrieving metadata
Loading datapackage
Creating metadata dataframes
Building merged dataframe
Compressing DataFrame


target_id  source_id  edge_amount     edge_type  \
0              5541       5541     1.000000    production   
1              5542       5542     0.032928  technosphere   
2              5542       5542     1.000000    production   
3              5548       5542     0.000567  technosphere   
4              5549       5542     0.005161  technosphere   
...             ...        ...          ...           ...   
39010868      13445      13445     1.000000    production   
39010869      13447      13447     1.000000    production   
39010870      13459      13459     1.000000    production   
39010871      13478      13478     1.000000    production   
39010872      13527      13527     1.000000    production   

                       target_database  \
0         EXIOBASE 3.8.1 2022 monetary   
1         EXIOBASE 3.8.1 2022 monetary   
2         EXIOBASE 3.8.1 2022 monetary   
3         EXIOBASE 3.8.1 2022 monetary   
4         EXIOBASE 3.8.1 2022 monetary   
...                                ...   
39010868  EXIOBASE 3.8.1 2022 monetary   
39010869  EXIOBASE 3.8.1 2022 monetary   
39010870  EXIOBASE 3.8.1 2022 monetary   
39010871  EXIOBASE 3.8.1 2022 monetary   
39010872  EXIOBASE 3.8.1 2022 monetary   

                                                target_code  \
0                              Cultivation of paddy rice|AT   
1                                   Cultivation of wheat|AT   
2                                   Cultivation of wheat|AT   
3                               Cultivation of crops nec|AT   
4                                         Cattle farming|AT   
...                                                     ...   
39010868  Re-processing of secondary copper into new cop...   
39010869  Re-processing of secondary other non-ferrous m...   
39010870            Recycling of bottles by direct reuse|WM   
39010871  Re-processing of secondary construction materi...   
39010872      Extra-territorial organizations and bodies|WM   

                                                target_name target_location  \
0                                 Cultivation of paddy rice              AT   
1                                      Cultivation of wheat              AT   
2                                      Cultivation of wheat              AT   
3                                  Cultivation of crops nec              AT   
4                                            Cattle farming              AT   
...                                                     ...             ...   
39010868  Re-processing of secondary copper into new copper              WM   
39010869  Re-processing of secondary other non-ferrous m...              WM   
39010870               Recycling of bottles by direct reuse              WM   
39010871  Re-processing of secondary construction materi...              WM   
39010872         Extra-territorial organizations and bodies              WM   

         target_unit target_type  \
0          million €     process   
1          million €     process   
2          million €     process   
3          million €     process   
4          million €     process   
...              ...         ...   
39010868   million €     process   
39010869   million €     process   
39010870   million €     process   
39010871   million €     process   
39010872   million €     process   

                                   target_reference_product  \
0                                 Cultivation of paddy rice   
1                                      Cultivation of wheat   
2                                      Cultivation of wheat   
3                                  Cultivation of crops nec   
4                                            Cattle farming   
...                                                     ...   
39010868  Re-processing of secondary copper into new copper   
39010869  Re-processing of secondary other non-ferrous m...   
39010870               Recycling of bottles by direct reuse   
390

And we can do the same thing for LCA matrices:

In [36]:
act = bd.Database("EXIOBASE 3.8.1 2022 monetary").random()
act

'Distribution and trade of electricity' (million €, CZ, None)

In [41]:
lca = bc.LCA({act: 1}, method=('IPCC 2013', 'climate change', 'GWP 100a'))
lca.lci()
lca.lcia()
lca.score

771164.4127744769

In [42]:
lca.to_dataframe()

row_index  col_index         amount  row_id  col_id row_database  \
0           21       1875  250227.301624    1229    6303   biosphere3   
1           37       1875     467.484767    1737    6303   biosphere3   
2           29       1875    5373.530425    1373    6303   biosphere3   
3           23       1875     226.716553    1241    6303   biosphere3   
4           21       1860  144437.577992    1229    6288   biosphere3   
..         ...        ...            ...     ...     ...          ...   
195         35       1773     160.954083    1691    6201   biosphere3   
196         35       1778     156.990175    1691    6206   biosphere3   
197         35       8624     130.299857    1691   13052   biosphere3   
198         35       6994     129.783952    1691   11422   biosphere3   
199         28       1767     146.459623    1371    6195   biosphere3   

                                 row_code                 row_name  \
0    349b29d1-3e58-4c66-98b9-9d1a076efd2e   Carbon dioxide, fossil   
1    0795345f-c7ae-410c-ad25-1845784c75f5          Methane, fossil   
2    20185046-64bb-4c09-a8e7-e8a9e144ca98      Dinitrogen monoxide   
3    ba2f3f82-c93a-47a5-822a-37ec97495275  Carbon monoxide, fossil   
4    349b29d1-3e58-4c66-98b9-9d1a076efd2e   Carbon dioxide, fossil   
..                                    ...                      ...   
195  6c977009-5c4e-4901-a4c1-ab20389cb972      Methane, non-fossil   
196  6c977009-5c4e-4901-a4c1-ab20389cb972      Methane, non-fossil   
197  6c977009-5c4e-4901-a4c1-ab20389cb972      Methane, non-fossil   
198  6c977009-5c4e-4901-a4c1-ab20389cb972      Methane, non-fossil   
199  81305cda-63f1-4c1a-99c6-2c2568550ec4      Dinitrogen monoxide   

    row_location  row_unit  row_type                          row_categories  \
0           None  kilogram  emission                                     air   
1           None  kilogram  emission                                     air   
2           None  kilogram  emission                                     air   
3           None  kilogram  emission                                     air   
4           None  kilogram  emission                                     air   
..           ...       ...       ...                                     ...   
195         None  kilogram  emission  air::low population density, long-term   
196         None  kilogram  emission  air::low population density, long-term   
197         None  kilogram  emission  air::low population density, long-term   
198         None  kilogram  emission  air::low population density, long-term   
199         None  kilogram  emission  air::low population density, long-term   

    row_product                  col_database  \
0          None  EXIOBASE 3.8.1 2022 monetary   
1          None  EXIOBASE 3.8.1 2022 monetary   
2          None  EXIOBASE 3.8.1 2022 monetary   
3          None  EXIOBASE 3.8.1 2022 monetary   
4          None  EXIOBASE 3.8.1 2022 monetary   
..          ...                           ...   
195        None  EXIOBASE 3.8.1 2022 monetary   
196        None  EXIOBASE 3.8.1 2022 monetary   
197        None  EXIOBASE 3.8.1 2022 monetary   
198        None  EXIOBASE 3.8.1 2022 monetary   
199        None  EXIOBASE 3.8.1 2022 monetary   

                                 col_code                           col_name  \
0           Steam and hot water supply|CZ         Steam and hot water supply   
1           Steam and hot water supply|CZ         Steam and hot water supply   
2           Steam and hot water supply|CZ         Steam and hot water supply   
3           Steam and hot water supply|CZ         Steam and hot water supply   
4    Production of electricity by coal|CZ  Production of electricity by coal   
..                                    ...                                ...   
195                     Cattle farming|CZ                     Cattle farming   
196                           Raw milk|CZ                           Raw milk   
197              

Filtering LCA matrices by default takes a maximum number of values to return, but can also be configured to return all values whose absolute value is above a certain fraction of the total score. In that case, the `cutoff` should be between zero and one.

In [44]:
lca.to_dataframe(cutoff=0.01, cutoff_mode="fraction")

row_index  col_index         amount  row_id  col_id row_database  \
0          21       1875  250227.301624    1229    6303   biosphere3   
1          21       1860  144437.577992    1229    6288   biosphere3   
2          21       7002   63153.868488    1229   11430   biosphere3   
3          37       7002   54524.151057    1737   11430   biosphere3   
4          21       1784   21887.131730    1229    6212   biosphere3   
5          36       1784    8856.675400    1734    6212   biosphere3   
6          21       1864   18390.216445    1229    6292   biosphere3   
7          21       1885   18001.995079    1229    6313   biosphere3   
8          37       1785   12786.322474    1737    6213   biosphere3   
9          37       7001   10154.086262    1737   11429   biosphere3   
10         37       1873   10067.285965    1737    6301   biosphere3   

                                row_code                row_name row_location  \
0   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
1   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
2   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
3   0795345f-c7ae-410c-ad25-1845784c75f5         Methane, fossil         None   
4   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
5   70ef743b-3ed5-4a6d-b192-fb6d62378555         Methane, fossil         None   
6   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
7   349b29d1-3e58-4c66-98b9-9d1a076efd2e  Carbon dioxide, fossil         None   
8   0795345f-c7ae-410c-ad25-1845784c75f5         Methane, fossil         None   
9   0795345f-c7ae-410c-ad25-1845784c75f5         Methane, fossil         None   
10  0795345f-c7ae-410c-ad25-1845784c75f5         Methane, fossil         None   

    row_unit  row_type                          row_categories row_product  \
0   kilogram  emission                                     air        None   
1   kilogram  emission                                     air        None   
2   kilogram  emission                                     air        None   
3   kilogram  emission                                     air        None   
4   kilogram  emission                                     air        None   
5   kilogram  emission  air::non-urban air or from high stacks        None   
6   kilogram  emission                                     air        None   
7   kilogram  emission                                     air        None   
8   kilogram  emission                                     air        None   
9   kilogram  emission                                     air        None   
10  kilogram  emission                                     air        None   

                    col_database  \
0   EXIOBASE 3.8.1 2022 monetary   
1   EXIOBASE 3.8.1 2022 monetary   
2   EXIOBASE 3.8.1 2022 monetary   
3   EXIOBASE 3.8.1 2022 monetary   
4   EXIOBASE 3.8.1 2022 monetary   
5   EXIOBASE 3.8.1 2022 monetary   
6   EXIOBASE 3.8.1 2022 monetary   
7   EXIOBASE 3.8.1 2022 monetary   
8   EXIOBASE 3.8.1 2022 monetary   
9   EXIOBASE 3.8.1 2022 monetary   
10  EXIOBASE 3.8.1 2022 monetary   

                                             col_code  \
0                       Steam and hot water supply|CZ   
1                Production of electricity by coal|CZ   
2   Extraction of natural gas and services related...   
3   Extraction of natural gas and services related...   
4   Mining of coal and lignite; extraction of peat|CZ   
5   Mining of coal and lignite; extraction of peat|CZ   
6                Production of electricity by wind|CZ   
7                             Other land transport|CZ   
8   Extraction of crude petroleum and services rel...   
9   Extraction of crude petroleum and services rel...   
10           Distribution and trade of electricity|CZ   

                                             col_name col_location   col_unit  \
0                          Ste